### PR TITLE
latex, allow image dimensions using px unit (96px=1in)

### PR DESCRIPTION
**Pull request by jfbu**
New key `'pixelsize'` in `latex_elements` dictionary. This works with same interface in all engines (pdflatex, lualatex, xelatex and hopefully platex which I can't test ;-) )

### CODE REVIEW

File: sphinx/writers/latex.py [hunk 1]
```diff
@@ -1380,9 +1384,10 @@ def latex_image_length(self, width_str):
             return width_str
         res = width_str
         amount, unit = match.groups()[:2]
-        if not unit or unit == "px":
-            # pixels: let LaTeX alone
+        if not unit:
             return None
+        elif unit == "px":
+            res = "%.3f\\sphinxpxdimen" % (float(amount))
```
Comment (jfbu) [inline sphinx/writers/latex.py:1390]: here I hesitated about format specifier, but `.3f` looks ok

**Comment by tk0miya** [comment]
I confirmed this works with pLaTeX.

Note: dvipdfmx expects DPI of images is 72. so I need to set `1bp` to `latex_elements['pixelsize']` .

**Comment by tk0miya** [comment]
+1 for merge!

**Comment by jfbu** [comment]
Indeed both `dvipdfmx` (via `extractbb`)  and `pdftex` use by default 72dpi for determining image dimensions from the number of pixels, when the image file doesn't contain information about the resolution. It seems `dvipdfmx`'s default is not user modifiable, but `pdftex` has `\pdfimageresolution` which is an integer, default `72` (texdoc pdftex). In my brief testing, my jpg and png files already had resolution information in the file thus I could not test impact of `\pdfimageresolution` but anyhow what I conclude from the doc of pdftex is that this is independent matter from the setting of `\pdfpxdimen`:

> The default value of `\pdfpxdimen` is 1 bp, corresponding to a pixel density of 72 dpi. This primitive
> is completely independent from the `\pdfimageresolution` and `\pdfpkresolution` parameters

Thus current PR does not seem to introduce difference between `dvipdfmx` and `pdftex` but the name `pixelsize` may lead to ambiguity. The key only has impact on interpretation of `width`, or `height` when given explicitely using `px` unit. It doesn't relate to how an image file with no resolution information is handled.

**Comment by jfbu** [comment]
perhaps `'pxunit'` would be more clear than `'pixelsize'` ?

**Comment by jfbu** [comment]
@tk0miya there is a problem in `latex.py`: this PR modifies `latex_image_length` from `LaTeXTranslator` class which is called only to handle `width` and `height` image directive attributes; but there is also `width_to_latex_length` which is called for `width` (`figwidth`) of figure directive. Even before PR the two differ because the latter silently replace `pt` by `bp`, but not `latex_image_length`. Hence `:width: 100pt` for image directive and `:figwidth: 100pt` for figure directive are not handled the same. This looks strange. Why two separate python functions  with diverging behaviour ?

Anyway, I am updating this PR to use `'pxunit'` name for the key rather than `'pixelsize'`.

**Comment by jfbu** [comment]
@tk0miya in my short investigation I see no reason why `width_to_latex_length` and `LaTeXTranslator.latex_image_length` differ. I will raise a separate issue.

**Comment by jfbu** [comment]
@tk0miya Thanks for checking out with `platex`! (I think I can test too, but have to revert to TL2015, because my TL2016 install did not include Japanese ...). I am merging now see PR #2714 for extended proposal.

**Pull request merged**